### ISSUE TITLE
Preserve radix when pretty-printing numeric literal types

### ISSUE THREAD

**Issue by iliazeus** [microsoft/TypeScript#45550]
# Suggestion

## 🔍 Search Terms

numeric literal radix, non-decimal numeric literal types

List of keywords you searched for before creating this issue. Write them down here so that others can find this suggestion more easily and help provide feedback.

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

If a numeric literal type was defined explicitly with a non-decimal literal, it should be displayed as such in editor hovers and other places where it is pretty-printed.

## 📃 Motivating Example

Often, non-decimal literals are more-or-less self-documenting in a context:

```typescript
declare const MASK = 0b1011; // it is obvious which bits are set
// vs
declare const MASK = 11; // wat?
```

```typescript
declare const PERMISSIONS: 0o777; // unix file system permissions
// vs
declare const PERMISSIONS: 511; // wat?
```

```typescript
declare const ARGB = 0x00ffff00; // yellow pixel color
// vs
declare const ARGB = 16776960; // wat?
```

```typescript
declare const ASCII_CODE = 0x1b; // ASCII code for backspace
// vs
declare const ASCII_CODE = 27; // wat?
```

Ideally, the radix should carry over to the "autoincremented" enum members:

```typescript
enum AsciiLetters {
  A = 0x61, B, C, D, E, F,  /* ... */, Z,
}
// AsciiLetters.E: 0x65
```

## 💻 Use Cases

Bit masks. Unix file system permissions. RGB colors. Hexadecimal ASCII/Unicode code points.

**Comment by iliazeus** [comment]
Possibly related to https://github.com/microsoft/TypeScript/issues/42197.

**Comment by andrewbranch** [comment]
> it should be displayed as such in editor hovers and other places where it is pretty-printed

#42197 seems to just be about JS emit. You mention quick info; are you also concerned about JS emit and/or declaration emit?

**Comment by iliazeus** [comment]
I am concerned about declaration emit, for the same reasons I described above.

Funnily enough, when I test it on the playground, JS emit preserves the radix, but not the .D.TS emit:

https://www.typescriptlang.org/play?ts=4.4.0-beta#code/KYDwDg9gTgLgBAYwgOwM7wLIHkAiBROAXjgAYIB2SgbgFgAoIA

**Comment by Kingwl** [comment]
It's might be a breaking change. Might should behind a flag?

**Comment by iliazeus** [comment]
@Kingwl 

> It's might be a breaking change. Might should behind a flag?

As far as I understand, it should break nothing in the type system itself. `0xf` and `15` should still be essentially the same type, in the same sense as `type Foo = number` and `type Bar = number` are.

**Comment by Kingwl** [comment]
> it should break nothing in the type system itself.

Exactly. But the breaking change not only contains the type system but also the code emit. There might some users who using the AST of emitted js (or d.ts) (eg: https://github.com/HearTao/ts-creator/issues/18). And it's a issue that's been going on for a long time (At least at 3.3.3).  

And I'm not 100% sure even it's seems enough to be a `breaking change`. Might @rbuckton  or @sandersn has some comments.

(And I'm happy to work on this. It's also related with https://github.com/microsoft/TypeScript/issues/30041

**Comment by andrewbranch** [comment]
Thanks @Kingwl! Changes to emit that are detectable by AST analysis but not by runtime behavior are not considered breaking changes.
